### PR TITLE
Optimize completion

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
### First invocation 

Before

image

After

image

### Second invocation 

Before

image

After

image

### CODE REVIEW

File: src/fsharp/service/ServiceAssemblyContent.fs [hunk 1]
```diff
@@ -241,7 +262,7 @@ module AssemblyContentProvider =
                           processIdents (fullCompiledIdents |> String.concat ".") fullCompiledIdents)
                      |> Option.toList ])
 
-    let rec private traverseEntity contentType (parent: Parent) (entity: FSharpEntity) =
```
Comment (cartermp) [inline src/fsharp/service/ServiceAssemblyContent.fs]: Why remove `private`? Just curious.

File: src/fsharp/service/ServiceAssemblyContent.fs [hunk 2]
```diff
@@ -308,7 +329,7 @@ module AssemblyContentProvider =
         |> Seq.distinctBy (fun {FullName = fullName; CleanedIdents = cleanIdents} -> (fullName, cleanIdents))
         |> Seq.toList
 
-    let private getAssemblySignaturesContent contentType (assemblies: FSharpAssembly list) =
```
Comment (cartermp) [inline src/fsharp/service/ServiceAssemblyContent.fs]: Same as with `traverseEntity`

File: src/fsharp/service/ServiceDeclarationLists.fs [hunk 3]
```diff
@@ -650,22 +652,24 @@ type FSharpDeclarationListInfo(declarations: FSharpDeclarationListItem[], isForT
                             | Some _ -> displayName
                             | None -> Lexhelp.Keywords.QuoteIdentifierIfNeeded displayName
 
-                    let isAttribute = SymbolHelpers.IsAttribute infoReader item.Item
-                    
                     let cutAttributeSuffix (name: string) =
-                        if isAttributeApplicationContext && isAttribute && name <> "Attribute" && name.EndsWith "Attribute" then
```
Comment (cartermp) [inline src/fsharp/service/ServiceDeclarationLists.fs]: Wild that this call was a bottleneck. Good find.

**Comment by cartermp** [comment]
@dotnet-bot test this please

**Comment by vasily-kirichenko** [comment]
@cartermp it's not compilable at the moment.

File: src/fsharp/service/ServiceDeclarationLists.fs (hunk 3 above)
Comment (vasily-kirichenko) [reply src/fsharp/service/ServiceDeclarationLists.fs]: Accessibility rules by the signature file, those modifiers are left from the time when there is no one.

File: src/fsharp/service/ServiceAssemblyContent.fs (hunk 1 above)
Comment (TIHan) [reply src/fsharp/service/ServiceAssemblyContent.fs]: It's already private because the signature file doesn't expose it anyway.

File: src/fsharp/service/ServiceAssemblyContent.fs (hunk 2 above)
Comment (TIHan) [reply src/fsharp/service/ServiceAssemblyContent.fs]: What I commented above.

File: src/fsharp/service/ServiceDeclarationLists.fs (hunk 3 above)
Comment (forki) [reply src/fsharp/service/ServiceDeclarationLists.fs]: Cutattributesuffix is called at least twice. How about using lazy expression here to make sure isAttribute is calculated at most once?

Comment (smoothdeveloper) [reply src/fsharp/service/ServiceDeclarationLists.fs]: @forki, it is called twice but on potentially different values (see how name and nameInCode are initialized), although it is right that we could optimize some of this because we never want to cut attribute suffix for those cases:

* operator
* back ticked name

I think it should apply only to `name` but not nameInCode which is different only for the cases mentioned above.

Comment (forki) [reply src/fsharp/service/ServiceDeclarationLists.fs]: I'm talking about the `SymbolHelpers.IsAttribute infoReader item.Item` part. That one was the bottleneck and stays stable over both calls. That's why it was factored out to the top. I propose we factor it back to the top but make it lazy.

File: src/fsharp/symbols/SymbolHelpers.fs [hunk 4]
```diff
@@ -827,10 +828,11 @@ module internal SymbolHelpers =
          | Item.Types(it, [ty]) -> 
              g.suppressed_types 
              |> List.exists (fun supp -> 
-                if isAppTy g ty && isAppTy g (generalizedTyconRef supp) then 
+                let generalizedSupp = generalizedTyconRef supp
+                if isAppTy g ty && isAppTy g generalizedSupp then
```
Comment (forki) [inline src/fsharp/symbols/SymbolHelpers.fs]: Can we move `isAppTy g ty` further out so that we only check that once?

File: src/fsharp/service/ServiceAssemblyContent.fs [hunk 5]
```diff
@@ -186,14 +187,31 @@ type IAssemblyContentCache =
...
-    let private createEntity ns (parent: Parent) (entity: FSharpEntity) =
+    let unresolvedSymbol (topRequireQualifiedAccessParent: Idents option) (cleanedIdents: Idents) (fullName: string) =
+        let getNamespace (idents: Idents) = 
+            if idents.Length > 1 then Some idents.[..idents.Length - 2] else None
+
+        let ns = 
+            topRequireQualifiedAccessParent 
+            |> Option.bind getNamespace 
+            |> Option.orElseWith (fun () -> getNamespace cleanedIdents)
+            |> Option.defaultValue [||]
```
Comment (forki) [inline src/fsharp/service/ServiceAssemblyContent.fs]: Can we use defaultWith or something?

File: src/fsharp/service/ServiceAssemblyContent.fs [hunk 6]
```diff
@@ -208,21 +226,25 @@ module AssemblyContentProvider =
...
+              UnresolvedSymbol = unresolvedSymbol topRequireQualifiedAccessParent cleanIdents fullName
             })
 
-    let private traverseMemberFunctionAndValues ns (parent: Parent) (membersFunctionsAndValues: seq<FSharpMemberOrFunctionOrValue>) =
+    let traverseMemberFunctionAndValues ns (parent: Parent) (membersFunctionsAndValues: seq<FSharpMemberOrFunctionOrValue>) =
         membersFunctionsAndValues
         |> Seq.filter (fun x -> not x.IsInstanceMember && not x.IsPropertyGetterMethod && not x.IsPropertySetterMethod)
         |> Seq.collect (fun func ->
             let processIdents fullName idents = 
+                let topRequireQualifiedAccessParent = parent.TopRequiresQualifiedAccess true |> Option.map parent.FixParentModuleSuffix
```
Comment (forki) [inline src/fsharp/service/ServiceAssemblyContent.fs]: This looks like we can move that further out

File: src/fsharp/service/ServiceAssemblyContent.fs (hunk 5 above)
Comment (vasily-kirichenko) [reply src/fsharp/service/ServiceAssemblyContent.fs]: What do you mean? `[||]` is a singleton, no allocations here.

**Comment by forki** [comment]
Never mind.

Vasily Kirichenko <notifications@github.com> schrieb am Sa., 2. Juni 2018,
09:11:

File: src/fsharp/service/ServiceAssemblyContent.fs [hunk 7]
```diff
@@ -208,21 +226,25 @@ module AssemblyContentProvider =
...
+                let cleanedIdentes = parent.FixParentModuleSuffix idents
                 { FullName = fullName
-                  CleanedIdents = parent.FixParentModuleSuffix idents
+                  CleanedIdents = cleanedIdentes
                   Namespace = ns
                   NearestRequireQualifiedAccessParent = parent.ThisRequiresQualifiedAccess true |> Option.map parent.FixParentModuleSuffix
-                  TopRequireQualifiedAccessParent = parent.TopRequiresQualifiedAccess true |> Option.map parent.FixParentModuleSuffix
+                  TopRequireQualifiedAccessParent = topRequireQualifiedAccessParent
                   AutoOpenParent = parent.AutoOpen |> Option.map parent.FixParentModuleSuffix
```
Comment (smoothdeveloper) [inline src/fsharp/service/ServiceAssemblyContent.fs]: This one is also candidate for moving out.

Although by moving both out the cost will be there even if the (filtered) sequence is empty.

**Comment by vasily-kirichenko** [comment]
@forki @smoothdeveloper all done, thanks.

**Comment by vasily-kirichenko** [comment]
Before caching ILTypeDef.CustomAttrs

image

After 

image

File: src/absil/il.fs [hunk 8]
```diff
@@ -1933,6 +1933,8 @@ type ILTypeDef(name: string, attributes: TypeAttributes, layout: ILTypeDefLayout
                extends: ILType option, methods: ILMethodDefs, nestedTypes: ILTypeDefs, fields: ILFieldDefs, methodImpls: ILMethodImplDefs,
                events: ILEventDefs, properties: ILPropertyDefs, securityDeclsStored: ILSecurityDeclsStored, customAttrsStored: ILAttributesStored, metadataIndex: int32) =  
 
+    let customAttrs = lazy (customAttrsStored.GetCustomAttrs metadataIndex)
```
[the remark this answers is not in the document]
Comment (dsyme) [reply src/absil/il.fs]: If you're going to store these, I suppose it should be `let mutable customAttrsStored = customAttrsStored` and then on access of `CustomAttrs` you replace any `ILAttributesStored.Reader` by `ILAttributesStored.Given` after evaluation.  That won't increase memory size of `ILTypeDef` at all

Thinking it over I think it is now ok to store them for `ILTypeDef`, since memory pressure has been reduced in VS.

[the remark this answers is not in the document]
Comment (vasily-kirichenko) [reply src/absil/il.fs]: @dsyme Thank you, done.

**Comment by dsyme** [comment]
This is great work!

**Comment by vasily-kirichenko** [comment]
Before storing attrs in ILTypeDef (after executing completion in FSharp.Editor project):

image

After:

image

**Comment by vasily-kirichenko** [comment]
@dsyme Do I understand that we still cannot use struct tuples in FSharp.Compiler.Private project?

File: src/absil/il.fs (hunk 8 above)
[the remark this answers is not in the document]
Comment (smoothdeveloper) [reply src/absil/il.fs]: @dsyme 

> At some point we moved to deliberately not persist the custom attributes because there were so many of them hanging around.

Is there a reason we need to persist all of them versus only storing those that matters to the compiler (only a known set of attribute types are relevant)?

> Now memory pressure is relieved a bit in VS we might be able to persist them
> Thinking it over I think it is now ok to store them for ILTypeDef, since memory pressure has been reduced in VS.

So long FCS is in process, that memory pressure remains a problem, I'd not assume that some of the other optimizations make it ok to just go back on storing all attributes at all times (after first call, but the call is probably done in many cases).

Can we devise a plan to have them tucked away when memory pressure situation occurs?

Can we look at a solution to only store attributes that are relevant to the compilation? this could be done in `ILBinaryReader.customAttrsReader` AFAIU: https://github.com/Microsoft/visualfsharp/blob/cc8bc2b4077711a2efc454edaae2b702f9984e1f/src/absil/ilread.fs#L2554-L2563

**Comment by vasily-kirichenko** [comment]
I spot a lack of fast comparer for `bool` values. This is number of allocations caused by a single completion popup on FSharp.Editor project:

image

After providing a comparer, it look like this:

image

It's eliminated 1M boxed bool and 700K tuples allocations.

**Comment by dsyme** [comment]
> Do I understand right that we still cannot use struct tuples in FSharp.Compiler.Private project?

You can if you really need to but can cause some problems with packaging various releases and the Fable build of the compiler - but it's normally clearer to introduce a named struct

**Comment by vasily-kirichenko** [comment]
@dsyme done https://github.com/Microsoft/visualfsharp/pull/5076

File: src/fsharp/service/ServiceAssemblyContent.fs (hunk 5 above)
Comment (dsyme) [reply src/fsharp/service/ServiceAssemblyContent.fs]: It's not a singleton
```
> System.Object.ReferenceEquals([| |], [| |]);;
val it : bool = false
```
If you want to add a singleton table to array.fs `Array.empty` that would make sense, but make it a separate PR.  We'd also have to consider if `[| |]` should use that table.

File: src/fsharp/service/ServiceAssemblyContent.fs [hunk 9]
```diff
@@ -186,14 +187,31 @@ type IAssemblyContentCache =
...
+        let getNamespace (idents: Idents) = 
+            if idents.Length > 1 then Some idents.[..idents.Length - 2] else None
+
+        let ns = 
+            topRequireQualifiedAccessParent 
+            |> Option.bind getNamespace 
+            |> Option.orElseWith (fun () -> getNamespace cleanedIdents)
+            |> Option.defaultValue [||]
+
+        let displayName = cleanedIdents |> Array.skip ns.Length |> String.concat "."
```
Comment (dsyme) [inline src/fsharp/service/ServiceAssemblyContent.fs]: How do you know this `skip` isn't skipping past the end of the idents?  Perhaps be defensive?

File: src/fsharp/service/ServiceAssemblyContent.fs (hunk 5 above)
Comment (vasily-kirichenko) [reply src/fsharp/service/ServiceAssemblyContent.fs]: Hmm, interesting. I'll think about the table. Do you mean something like `Dictionary<Type, Type[]>`? Would lookup be 100% faster than allocating an empty array?

File: src/fsharp/service/ServiceAssemblyContent.fs (hunk 9 above)
Comment (vasily-kirichenko) [reply src/fsharp/service/ServiceAssemblyContent.fs]: Done.

**Comment by vasily-kirichenko** [comment]
> You can if you really need to but can cause some problems with packaging various releases and the Fable build of the compiler - but it's normally clearer to introduce a named struct

OK, replacing tuples with structs needs benchmarking anyway.

**Comment by dsyme** [comment]
Re Array.Empty, I meant:
```
type EmptyArray<'T>() = 
    static let empty = ([| |] : 'T [])
    static member Empty = empty
```
and so on. i.e. type-indexed by automagic of .NET generics.

.NET 4.something actually has `Array.Empty()` here:
```
> (System.Array.Empty() : int[]);;
val it : int [] = [||]
```
which is a singleton.  Hwoever it's only available since .NET 4.6 and FSharp.Core needs to run on .NET 4.5.  We could use it for .NET Standard though

File: src/fsharp/service/ServiceAssemblyContent.fs (hunk 5 above)
Comment (dsyme) [reply src/fsharp/service/ServiceAssemblyContent.fs]: answered in main comment section

**Comment by vasily-kirichenko** [comment]
@dsyme I've added the array empty singleton.

File: src/fsharp/service/ServiceDeclarationLists.fs [hunk 10]
```diff
@@ -676,7 +682,7 @@ type FSharpDeclarationListInfo(declarations: FSharpDeclarationListItem[], isForT
                             | None -> ns)
                         |> Option.bind (function
                             | [||] -> None
-                            | ns -> Some (ns |> String.concat "."))
+                            | ns -> Some (System.String.Join(".", ns)))
```
Comment (smoothdeveloper) [inline src/fsharp/service/ServiceDeclarationLists.fs]: Any reason String.Join is prefered over String.concat?

**Comment by vasily-kirichenko** [comment]
So, this makes completion 2x faster on FSharp.Editor project: 

### Master 

image

### This branch (release)

image

**Comment by vasily-kirichenko** [comment]
It turns out that

```fsharp
open System
open BenchmarkDotNet
open BenchmarkDotNet.Attributes
open BenchmarkDotNet.Running

type Test() =
    [<Benchmark(Baseline = true)>]
    member __.Sprintf() = sprintf "%06d" 123456

    [<Benchmark>]
    member __.Format() = (123456).ToString("D6")

[<EntryPoint>]
let main _ = 
    BenchmarkRunner.Run<Test>() |> ignore
    0
```

```
BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i5-7600 CPU 3.50GHz (Kaby Lake), 1 CPU, 4 logical and 4 physical cores
  [Host]     : .NET Framework 4.6.1 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.3101.0
  DefaultJob : .NET Framework 4.6.1 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.3101.0


  Method |     Mean |    Error |   StdDev | Scaled |
-------- |---------:|---------:|---------:|-------:|
 Sprintf | 340.3 ns | 5.908 ns | 5.237 ns |   1.00 |
  Format | 111.1 ns | 1.295 ns | 1.148 ns |   0.33 |
```

File: src/fsharp/FSharp.Core/array.fs [hunk 11]
```diff
@@ -74,8 +74,12 @@ namespace Microsoft.FSharp.Collections
             if array.Length = 0 then invalidArg "array" (SR.GetString(SR.notEnoughElements))            
             Microsoft.FSharp.Primitives.Basics.Array.subUnchecked 1 (array.Length - 1) array
 
+        type EmptyArray<'T>() =
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fs]: Could you do this as a separate PR please?  It could have wide reaching impact and we might not take it in F# 4.5. Thanks

Comment (dsyme) [reply src/fsharp/FSharp.Core/array.fs]: In particular I don't really want a situation where people will use `Array.empty` instead of `[| |]` just to get the singleton.  So doing the work in F# compiler to make `[| |]` call `Array.empty` should be part of adding this.  And we should assess the overall perf impact too, and also make sure the functions like `Array.map` return the singleton if the input was empty,

Comment (dsyme) [reply src/fsharp/FSharp.Core/array.fs]: Doing it in a separate PR will allow us to address all these things at once (though in F# 5.0, not 4.5).  Or at least allows us to discuss that

File: src/fsharp/PrettyNaming.fs [hunk 12]
```diff
@@ -138,7 +138,7 @@ module public Microsoft.FSharp.Compiler.PrettyNaming
     let IsOperatorName (name: string) =
         let name = if name.StartsWith "( " && name.EndsWith " )" then name.[2..name.Length - 3] else name
         // there is single operator containing a space - range operator with step: `.. ..`
-        let res = name = ".. .." || name |> Seq.forall (fun c -> opCharSet.Contains c && c <> ' ')
+        let res = name = ".. .." || name |> Seq.forall (fun c -> c <> ' ' && opCharSet.Contains c)
```
Comment (smoothdeveloper) [inline src/fsharp/PrettyNaming.fs]: consider if code is more readable with pattern match on `name` for the constant pattern first:

```fsharp
let IsOperatorName (name: string) =
    let name = if name.StartsWith "( " && name.EndsWith " )" then name.[2..name.Length - 3] else name
    // there is single operator containing a space - range operator with step: `.. ..`
    match name with
    | ".. .." -> true
    | _ -> name |> Seq.forall (fun c -> c <> ' ' && opCharSet.Contains c)
```

Comment (vasily-kirichenko) [reply src/fsharp/PrettyNaming.fs]: I don't think your variant is more readable.

**Comment by cartermp** [comment]
@dsyme can you re-review this? We have until Tuesday afternoon, then we will make the cutoff for dev15.8. I'd love to get this in for that.

File: vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs [hunk 13]
```diff
@@ -39,12 +39,13 @@ type internal FSharpCompletionProvider
     static let userOpName = "CompletionProvider"
     // Save the backing data in a cache, we need to save for at least the length of the completion session
     // See https://github.com/Microsoft/visualfsharp/issues/4714
-    static let declarationItemsData = new ConcurrentDictionary<string, FSharpDeclarationListItem>()
+    static let mutable declarationItems: FSharpDeclarationListItem[] = [||]
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs]: This is indeed much better
